File: homework1/src/warmup.js
// NUMBER ONE
/*
 * Returns an array with the minimum number of U.S. quarters, dimes, nickels, and
 * pennies, respectively, that make the given amount.
 */
function change(amount) {
    if (amount < 0) {
        throw new RangeError("amount cannot be negative");
    }
    const result = [];
    [25, 10, 5, 1].forEach((coin_value) => {
        result.push(Math.floor(amount / coin_value));
        amount %= coin_value;
    });
    return result;
}

// NUMBER TWO
/*
 * Returns a copy of the string with apostrophes and double quotes removed.
 */
function stripQuotes(s) {
    return s.replace(/['"]/g, '');
}

// NUMBER THREE
/*
 * Returns a random permutation of a string
 */
function scramble(s) {
    const array = s.split('');
    const n = s.length;

    for (let i = n - 1; i > 0; i--) {
        let j = Math.floor(Math.random() * (i + 1));
        let temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
    return array.join('');
}

// NUMBER FOUR
/*
 * Produces successive powers of a base, up to the given limit, passing each to a
 * callback.
 */
function powers(base, limit, callback) {
    let exponent = 0;
    while ((base ** exponent) <= limit) {
        callback(base ** exponent);
        exponent++;
    }
}

// NUMBER FIVE
/*
 * A generator that generates successive powers of a base, up to the given limit.
 */
function* powersGenerator(base, limit) {
    let value = 1;
    while (value <= limit) {
        yield value;
        value *= base;
    }
}

// NUMBER SIX
/*
 * Returns a string of words that were passed through the chainable
 * function, say, separated by spaces. The string is returned when nothing
 * is passed into say.
*/
function say(word) {
    if (!word) {
        return '';
    } else {
        return nextWord => !nextWord? word :
            say(word + ' ' + nextWord);
    }
}

// NUMBER SEVEN
/*
 * Returns the interleaving of an array with a bunch of values. The lengths of the
 * array and the number of values do not need to be the same.
 */
function interleave(array, ...values) {
    const result = [];
    const arrayLength = array.length;
    const valuesLength = values.length;
    const max = Math.max(arrayLength, valuesLength);
    for (let i = 0; i < max; i++) {
        if (i < arrayLength) {
            result.push(array[i]);
        }
        if (i < valuesLength) {
            result.push(values[i]);
        }
    }
    return result;
}

// NUMBER EIGHT
/*
 * Creates a cylinder object in the "Crockford Classes" style. There are no units for
 * the radius and height.
 */
function cylinder({radius = 1, height = 1}) {

    const surfaceArea = () => ((2 * Math.PI * radius * height) + (2 * Math.PI * radius * radius));
    const volume = () => (Math.PI * radius * radius * height);
    const widen = (widenFactor) => { radius *= widenFactor; };
    const stretch = (stretchFactor) => { height *= stretchFactor; };
    const toString = () => `Cylinder with radius of ${radius} and height of ${height}`;
    return Object.freeze({
        surfaceArea,
        volume,
        toString,
        get radius() {return radius;},
        get height() {return height;},
        widen,
        stretch
    });
}

// NUMBER NINE
/*
 * Returns an array of two functions, an encyptor and a decryptor, each using a
 * given key and a given encryption algorithm. The encryptor turns a UTF-8 encoded
 * string into a hex-string; the decryptor does the reverse.
 */
function makeCryptoFunctions(key, algorithm) {
    let crypto = require('crypto')
    const encrypt = function(toEnc) {
        let mykey = crypto.createCipher(algorithm, key)
        let mystr = mykey.update(toEnc, 'utf8', 'hex')
        mystr += mykey.final('hex')
        return mystr
    }
    const decrypt = function(toDec) {
        let mykey = crypto.createDecipher(algorithm, key)
        let mystr = mykey.update(toDec, 'hex', 'utf8')
        mystr += mykey.final('utf8')
        return mystr
    }
    return [encrypt, decrypt]
}

// NUMBER TEN
/*
 * Returns a promise for a name of the form "surname, name" obtained from the Uinames
 * API, for the given region and gender.
 */
function randomName({region, gender}) {
    const rp = require('request-promise');
    let options = {
        uri: 'https://uinames.com/api/',
        qs: {
            gender,
            region
        },
        json: true
    };
    return rp(options)
        .then(body => `${body.name}, ${body.surname}`);
}

module.exports = {
    change,
    stripQuotes,
    scramble,
    powers,
    powersGenerator,
    say,
    interleave,
    cylinder,
    makeCryptoFunctions,
    randomName
}

File: homework2/warmup.py
import math
from cryptography.fernet import Fernet
import re
import json
import requests
import random

#1
    """
    Returns a tuple with the minimum number of U.S. quarters, dimes, nickels,
    and pennies, respectively, that make the given amount.
    """
def change(number):
    if number < 0:
        raise ValueError('amount cannot be negative')

    quarters, remaining = divmod(amount, 25)
    dimes, remaining = divmod(remaining, 10)
    nickels, pennies = divmod(remaining, 5)
    return (quarters, dimes, nickels, pennies)

#2
    """
    Returns the string like s but with apostrophes and quotation marks removed.
    """
def strip_quotes(s):
    return re.sub('[\'"]', '', s)

#3
    """
    Return a random permutation of a string.
    """
def scramble(string):
    return ''.join(random.sample(string, len(string)))

#4
    """Returns a generator for powers of base up to limit"""
def powers(base, limit):
    value = 1
    while value <= limit:
        yield value
        value *= base

#5
    """
    Return a list of integer Pythagorean triples with max hypotenuse, limit.
    """
def triples(limit):
    return [(x, y, z)
        for x in range(1, limit + 1)
        for y in range(x, limit + 1)
        for z in range(y, limit + 1)
        if x * x + y * y == z * z]

#6
    """Returns a string of words from the famous chainable function problem."""
def say(a = ''):
    if a == '':
        return a
    def sayAgain(b = ''):
        if b == '':
            return a
        return say(a + ' ' + b)
    return sayAgain

#7
    """
    Returns the interleaving of an array with a bunch of values. The lengths
    of the array and the number of values do not need to be the same.
    """
def interleave(array, *argv):
    result = []
    arrayLength = len(array)
    valuesLength = 0
    for arg in argv:
        valuesLength+=1
    max_len = max(arrayLength, valuesLength)
    for i in range(max_len):
        if (i < arrayLength):
            result.append(array[i])
        if (i < valuesLength):
            result.append(argv[i])
    return result

#8
class Cylinder:
    def __init__(self, radius = 1, height = 1):
        self.radius = radius
        self.height = height
    @property
    def volume(self):
        return math.pi * self.radius * self.radius * self.height
    @property
    def surface_area(self):
        return 2 * math.pi * self.radius * self.height + (2 * math.pi * self.radius * self.radius)
    def stretch(self, stretch_factor):
        self.height *= stretch_factor
    def widen(self, widen_factor):
        self.radius *= widen_factor


#9
    """
    Return a tuple of two functions, one for encryption and one for decryption
    """
def make_crypto_functions(key):
    cipher = Fernet(key)
    return (cipher.encrypt, cipher.decrypt)

#10
    """
    Return a random name with the given region and geneder using the uinames API.
    """
def random_name(**k):
    url = 'http://uinames.com/api/'
    info = requests.get(url=url, params=k)
    person = json.loads(info.content)
    if 'error' in person:
        raise ValueError('{"error": "Invalid gender"}')
    result = person.get('surname') + ', ' + person.get('name')
    return result
